File: clients/php/Model/Player.php
<?php

namespace Model {
    require_once 'Model/Vehicle.php';
    require_once 'Stream.php';

    /**
     * Player (game participant)
     */
    class Player
    {
        /**
         * Index
         */
        public int $index;
        /**
         * Current score
         */
        public int $score;
        /**
         * List of player's vehicles
         */
        public array $vehicles;
    
        function __construct(int $index, int $score, array $vehicles)
        {
            $this->index = $index;
            $this->score = $score;
            $this->vehicles = $vehicles;
        }
    
        /**
         * Read Player from input stream
         */
        public static function readFrom(\InputStream $stream): Player
        {
            $index = $stream->readInt32();
            $score = $stream->readInt64();
            $vehicles = [];
            $vehiclesSize = $stream->readInt32();
            for ($vehiclesIndex = 0; $vehiclesIndex < $vehiclesSize; $vehiclesIndex++) {
                $vehiclesElement = \Model\Vehicle::readFrom($stream);
                $vehicles[] = $vehiclesElement;
            }
            return new Player($index, $score, $vehicles);
        }
        
        /**
         * Write Player to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeInt32($this->index);
            $stream->writeInt64($this->score);
            $stream->writeInt32(count($this->vehicles));
            foreach ($this->vehicles as $element) {
                $element->writeTo($stream);
            }
        }
    }
}

File: clients/cpp/model/Player.cpp
#include "Player.hpp"

namespace model {

Player::Player(int index, long long score, std::vector<model::Vehicle> vehicles) : index(index), score(score), vehicles(vehicles) { }

// Read Player from input stream
Player Player::readFrom(InputStream& stream) {
    int index = stream.readInt();
    long long score = stream.readLongLong();
    std::vector<model::Vehicle> vehicles = std::vector<model::Vehicle>();
    size_t vehiclesSize = stream.readInt();
    vehicles.reserve(vehiclesSize);
    for (size_t vehiclesIndex = 0; vehiclesIndex < vehiclesSize; vehiclesIndex++) {
        model::Vehicle vehiclesElement = model::Vehicle::readFrom(stream);
        vehicles.emplace_back(vehiclesElement);
    }
    return Player(index, score, vehicles);
}

// Write Player to output stream
void Player::writeTo(OutputStream& stream) const {
    stream.write(index);
    stream.write(score);
    stream.write((int)(vehicles.size()));
    for (const model::Vehicle& vehiclesElement : vehicles) {
        vehiclesElement.writeTo(stream);
    }
}

// Get string representation of Player
std::string Player::toString() const {
    std::stringstream ss;
    ss << "Player { ";
    ss << "index: ";
    ss << index;
    ss << ", ";
    ss << "score: ";
    ss << score;
    ss << ", ";
    ss << "vehicles: ";
    ss << "[ ";
    for (size_t vehiclesIndex = 0; vehiclesIndex < vehicles.size(); vehiclesIndex++) {
        const model::Vehicle& vehiclesElement = vehicles[vehiclesIndex];
        if (vehiclesIndex != 0) {
            ss << ", ";
        }
        ss << vehiclesElement.toString();
    }
    ss << " ]";
    ss << " }";
    return ss.str();
}

}